client/utils/guidance.py

```diff
@@ -0,0 +1,210 @@
+        """ 
+        Based on the student's answer (input_lines), we grab each associated
+        message if its corresponding misunderstanding's count is above the threshold
+        """    
+        if self.error:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        shorten_unique_id = assess_id_util.canon(unique_id)
+        # Try to get th info dictionary for this question
+        wa_2_dict_info = self.guidance_json['dictAssessId2WA2DictInfo'].get(shorten_unique_id)
+
+        if not wa_2_dict_info: # get returns None if wa_2_dict_info doesn't have shorten_unique_id in dictionary
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        dict_info = wa_2_dict_info.get(repr(input_lines))
+
+        # If this wrong answer is not in the JSON file, display default message
+        if not dict_info:            
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        self.set_tg(access_token,guidance_flag)
+        if self.tg_id == TG_ERROR_VALUE: 
+            # If self.tg_id == -1, some errors happen when trying to access server
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        lambda_string_key = self.guidance_json['dictTg2Func'].get(str(self.tg_id))
+
+        if not lambda_string_key:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        lambda_info_misu = lambda_string_key_to_func.get(lambda_string_key)
+        if not lambda_info_misu:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        lst_mis_u = dict_info.get('lstMisU')
+
+        # No list of misunderstandings for this wrong answer, default message
+        if not lst_mis_u:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        self.misU_count = self.update_misUcounts(hash_key, lst_mis_u, repr(input_lines),shorten_unique_id)
+
+        threshold = self.guidance_json['wrongAnsThresh']
+
+        self.msg_id_set = set()
+
+        # If the count is higher than the threshold, we need to display the message
+        for mis_u in self.misU_count:
+            if self.misU_count[mis_u] >= threshold:
+                # Add each associated misunderstanding ID to the set
+                msg_id = lambda_info_misu(dict_info, mis_u)
+
+                if msg_id:
+                    self.msg_id_set.add(msg_id)
+
+        if len(self.msg_id_set) == 0 or self.tg_id == TG_CONTROL:
+            # if student is in control group, just print the default message
+            print (DEFAULT_MSG)
+            return (self.misU_count, self.tg_id,"")
+
+        print("-- Helpful Hint --\n")
+
+        printed_out_msgs = ""
+        for mid in self.msg_id_set:
+            msg = self.guidance_json['dictId2Msg'][str(mid)]
+            printed_out_msgs = printed_out_msgs + msg
+            print(msg)
+
+        print("\n-- Helpful Hint --")
```

Comment:
I agree with Kyle that this can be confusing to students. Consider changing the messages to 

```python
print("-- Begin helpful hint --\n")
...
print("-- End helpful hint --")
```

If you really want to leave it like this, please add a comment in the code explaining that the second print is for bracketing purposes.